pkg/infinispan/grafana/dashboard_controller.go

```diff
@@ -0,0 +1,103 @@
+package grafana
+
+import (
+	"context"
+
+	grafanav1alpha1 "github.com/integr8ly/grafana-operator/v3/pkg/apis/integreatly/v1alpha1"
+	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/discovery"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("infinispan-grafana")
+
+func AddToScheme(scheme *runtime.Scheme) error {
+	return grafanav1alpha1.AddToScheme(scheme)
+}
+
+func EnsureDashboardExported(context context.Context, namespace string, mgr manager.Manager) error {
+	dc, err := discovery.NewDiscoveryClientForConfig(mgr.GetConfig())
+	if err != nil {
+		return err
+	}
+	// detect the GrafanaDashboard resource type resourceExists on the cluster
+	resourceExists, _ := k8sutil.ResourceExists(dc, grafanav1alpha1.SchemeGroupVersion.String(), grafanav1alpha1.GrafanaDashboardKind)
+	if !resourceExists {
+		log.Info("Grafana CRD not present - not installing dashboard CR")
+		return nil
+	}
+
+	infinispanDashboard := &grafanav1alpha1.GrafanaDashboard{}
+
+	selector := client.ObjectKey{
+		Name:      ApplicationName,
+		Namespace: namespace,
+	}
+
+	// Using mgr.GetClient() does not work as we don't configure the cache for this scheme
+	controllerClient, err := client.New(mgr.GetConfig(), client.Options{})
+	if err != nil {
+		return err
+	}
+
+	log.Info("Checking if grafana dashboard already exists")
+	err = controllerClient.Get(context, selector, infinispanDashboard)
+
+	if err == nil {
+		log.Info("Grafana dashboard already exists, not creating")
+		return nil
+	}
+	// We verify that the error is not a NotFound - if it is NotFound we ignore so we can create the dashboard
+	if !meta.IsNoMatchError(err) && !errors.IsNotFound(err) {
+		log.Error(err, "Error encountered while retrieving grafana dashboard")
+		return err
+	}
+
+	infinispanDashboard = dashboard(namespace)
+
+	log.Info("Creating grafana dashboard in namespace %s\n", namespace)
+	err = controllerClient.Create(context, infinispanDashboard)
+	if err != nil {
+		log.Error(err, "Error encountered while creating grafana dashboard\n")
+		return err
+	}
+
+	log.Info("Successfully grafana dashboard")
+	return nil
+}
```

Comment:
Following from  our Zulip conversation: `CreateOrPatch` can be used here. The "status" field that's expected in the struct is not an issue, as the `GrafanaDashboard` struct has the `Status GrafanaDashboardStatus` field. This is expected for CR structs and you will find it on most (all?) Kubernete CRs.

Reply:
I wasn't able to use `CreateOrPatch` as the status field is a different type (complained about not being able to convert to uint64 iirc).

Reply:
The error is the following
```
panic: cannot deep copy uint64

goroutine 1 [running]:
k8s.io/apimachinery/pkg/runtime.DeepCopyJSONValue(0x1440f00, 0x2099020, 0xc000aada38, 0xc0003976e0)
	/home/wburns/RedHat/external/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/runtime/converter.go:474 +0x4e6
k8s.io/apimachinery/pkg/runtime.DeepCopyJSONValue(0x149ab20, 0xc000397470, 0x1, 0x1)
	/home/wburns/RedHat/external/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/runtime/converter.go:458 +0x157
k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.NestedFieldCopy(0xc0003960c0, 0xc000aadc08, 0x1, 0x1, 0x0, 0x0, 0xa, 0x17e4300, 0xc0002f2780)
	/home/wburns/RedHat/external/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/apis/meta/v1/unstructured/helpers.go:43 +0xa6
github.com/infinispan/infinispan-operator/pkg/kubernetes.CreateOrPatch(0x180f3a0, 0xc000044140, 0x181d620, 0xc0003fe340, 0x17e4300, 0xc0002f2780, 0x0, 0x0, 0x0, 0xc00091fcb8, ...)
	/home/wburns/RedHat/infinispan-operator/pkg/kubernetes/controllerutil.go:63 +0x4bf
github.com/infinispan/infinispan-operator/pkg/infinispan/grafana.Create(0xc0000424d0, 0xa, 0x18245e0, 0xc00048bc80, 0x0, 0x0)
	/home/wburns/RedHat/infinispan-operator/pkg/infinispan/grafana/dashboard_controller.go:46 +0x1c6
github.com/infinispan/infinispan-operator/pkg/launcher.Launch(0x0)
	/home/wburns/RedHat/infinispan-operator/pkg/launcher/launcher.go:145 +0x4f3
main.main()
	/home/wburns/RedHat/infinispan-operator/cmd/manager/main.go:8 +0x2a
Process exiting with code: 2 signal: false
```

Reply:
> Which happens if the dashboard is already present when using CreateOrPatch. From the line number it is attempted to parse out the JSON and a type doesn't match the expected one.

You should provide different code for Create and Patch via mutate function to update only reflected sections or skip all updates

Reply:
Not in this code, this is for CreateOrPatch which Ryan suggested

Reply:
That is the code I am talking about if you look at the stack trace. I replaced my code with `CreateOrPatch` and I get the error and the mutate function is never referenced.

```
github.com/infinispan/infinispan-operator/pkg/kubernetes.CreateOrPatch(0x180f3a0, 0xc000044140, 0x181d620, 0xc0003fe340, 0x17e4300, 0xc0002f2780, 0x0, 0x0, 0x0, 0xc00091fcb8, ...)
	/home/wburns/RedHat/infinispan-operator/pkg/kubernetes/controllerutil.go:63 +0x4bf
```

Reply:
@wburns could you please to share updated code

Reply:
@dmvolod I added an additional commit that uses `CreateAndPatch`. It works great if the dashboard isn't present, but it if is then it blows up with the above stack trace.

Reply:
Here is a gist of the output from the command line as well and log for the commit I am running.
https://gist.github.com/wburns/a632a622f02673f147085f91ff471054

Reply:
Changing it to `CreateOrUpdate` seems to work.